File: PLScala/MergeSort.scala
import java.util.Random


object MergeSort extends App {

  var size = 1000
  var array1 = new Array[Int](size)

  var x = new Random()

  for(i <- (0 to size-1)){
      array1(i) = x.nextInt(1000)
  }

  var result = new Array[Int](1000)

  val startTime = System.currentTimeMillis();
  result = mergeSort(array1, 0, array1.length-1)
  val endTime = System.currentTimeMillis();
  val duration = endTime - startTime
  println(" Time taken for sequential : " + duration)



  def mergeArray(array1: Array[Int], l: Int, m: Int, r: Int) : Array[Int] = {

    var z = new Array[Int](1000)
    for(i <- (0 to size-1)){
      z(i) = array1(i)
    }

   var i = l
   var j = m + 1
   var k = l

    while ( {
      i <= m && j <= r
    }) {
      if (z(i) <= z(j)) {
        array1(k) = z(i)
        i += 1
      }
      else {
        array1(k) = z(j)
        j += 1
      }
      k += 1
    }
    // Copy the rest of the left side of the array into the target array
    while ( {
      i <= m
    }) {
      array1(k) = z(i)
      k += 1
      i += 1
    }

    z
  }

  def mergeSort(array1: Array[Int], l: Int, r: Int) : Array[Int] ={

    var y = new Array[Int](1000)

    if(l < r){
      var m = (l+r)/2
      mergeSort(array1, l, m)
      mergeSort(array1, m+1, r)

      y = mergeArray(array1, l, m, r)
    }
    y
  }





}

File: SortingAlgorithms/src/MergeSort.java
import java.util.Random;

public class MergeSort {


    private static final int LIST_SIZE = 1000000;
    private static final int MAX_VALUE = 1000;

    public static void main(String args[]) {

        Integer[] result;
        MergeSort sort = new MergeSort();

        Integer[] items = new Integer[LIST_SIZE];
        Random rand = new Random();
        for(int i = 0; i < LIST_SIZE; i++) {
            items[i] =rand.nextInt() % MAX_VALUE;
        }

        long startTime = System.currentTimeMillis();
        sort.mergeSort(items);
        sort.mergeSort(items);
        sort.mergeSort(items);
        long endTime = System.currentTimeMillis();


        System.out.println("Time Taken Sequential :: "+ (endTime - startTime));


    }

    public Integer[] mergeSort(Integer[] items) {
        if(items.length == 1)
            return items;
        //System.out.println("Inside");
        int splitPoint = items.length / 2;
        Integer[] leftArray = new Integer[splitPoint];
        Integer[] rightArray = new Integer[items.length - splitPoint];
        int traverseIndex = 0;
        for(int i = 0; i < splitPoint; i++) {
            leftArray[i] = items[traverseIndex];
            traverseIndex++;
        }

        for(int i = 0; i < (items.length - splitPoint); i++) {
            rightArray[i] = items[traverseIndex];
            traverseIndex++;
        }

        return merge(mergeSort(leftArray), mergeSort(rightArray));
    }

    private Integer[] merge(Integer[] left, Integer[] right) {
        int leftIndex = 0;
        int rightIndex = 0;
        Integer[] merged = new Integer[left.length + right.length];
        int mergedIndex = 0;

        while (leftIndex < left.length && rightIndex < right.length) {
            if (left[leftIndex] <= right[rightIndex]) {
                merged[mergedIndex] = left[leftIndex];
                leftIndex += 1;
            } else {
                merged[mergedIndex] = right[rightIndex];
                rightIndex += 1;
            }
            mergedIndex++;
        }
        if (leftIndex == left.length) {
            for(int i = rightIndex; i < right.length; i++) {
                merged[mergedIndex] = right[i];
                mergedIndex++;
            }
        } else {
            for(int i = leftIndex; i < left.length; i++) {
                merged[mergedIndex] = left[i];
                mergedIndex++;
            }
        }
        return merged;

    }

}
